### PR TITLE
[leeco] Add live.le.com functionality

### LINKED ISSUE #11788: [LeEco] Support Le.com live channel
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like that [x])
- Use *Preview* tab to see how your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2017.01.18*. If it's not read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [ ] I've **verified** and **I assure** that I'm running youtube-dl **2017.01.18**

Version: 2016.12.01

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through [README](https://github.com/rg3/youtube-dl/blob/master/README.md) and **most notably** [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [x] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):

轮播
* Live channel: http://live.le.com/lunbo/play/index.shtml?channel=225

### PR DESCRIPTION
**Pull request by einstein95**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds live.le.com functionality to the leeco.py extractor. This fixes #11788.

### CODE REVIEW

File: youtube_dl/extractor/leeco.py [hunk 1]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+            'ext': 'm3u8',
+            'title': '电视剧',
+            'is_live': True,
+        }
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        # webpage = self._download_webpage(url, video_id)
+        json = self._download_json('http://player.pc.le.com/player/startup_by_channel_id/1001/%s?host=live.le.com' % video_id, video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Use `query` for query.

File: youtube_dl/extractor/leeco.py [hunk 2]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+        streams = json['streams']
+        formats = []
+        for stream in streams:
+            surl = stream.get('streamUrl')
+            quality = stream.get('rate')
+            if surl:
+                surl = self._download_json(surl + '&format=1&expect=2&platid=10&splatid=1001&uuid=28394F488CFCF69F405F55B6B046EA4AAC32713C_0&vkit=20161017', video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Use `query` for query.

File: youtube_dl/extractor/leeco.py [hunk 3]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+                surl = surl['location']
+                formats.append({
+                    'url': surl,
+                    'ext': 'm3u8',
+                    'quality': quality
+                })
+        self._sort_formats(formats)
+
+        titles = self._download_json('http://static.api.letv.com/live/proxy?belongArea=100&clientId=1001', video_id)['data']
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Use `query` for query.

File: youtube_dl/extractor/leeco.py [hunk 4]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+        for entry in titles:
+            if entry['channelId'] == video_id:
+                title = entry['channelName']
+                break
+
+        return {
+            'id': video_id,
+            'title': title,
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: `_live_title` should be used.

File: youtube_dl/extractor/leeco.py [hunk 5]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+        formats = []
+        for stream in streams:
+            surl = stream.get('streamUrl')
+            quality = stream.get('rate')
+            if surl:
+                surl = self._download_json(surl + '&format=1&expect=2&platid=10&splatid=1001&uuid=28394F488CFCF69F405F55B6B046EA4AAC32713C_0&vkit=20161017', video_id)
+                surl = surl['location']
+                formats.append({
+                    'url': surl,
+                    'ext': 'm3u8',
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Should be `mp4`.

File: youtube_dl/extractor/leeco.py [hunk 6]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+        # webpage = self._download_webpage(url, video_id)
+        json = self._download_json('http://player.pc.le.com/player/startup_by_channel_id/1001/%s?host=live.le.com' % video_id, video_id)
+        streams = json['streams']
+        formats = []
+        for stream in streams:
+            surl = stream.get('streamUrl')
+            quality = stream.get('rate')
+            if surl:
+                surl = self._download_json(surl + '&format=1&expect=2&platid=10&splatid=1001&uuid=28394F488CFCF69F405F55B6B046EA4AAC32713C_0&vkit=20161017', video_id)
+                surl = surl['location']
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Should no break if some format has no location.

File: youtube_dl/extractor/leeco.py [hunk 7]
```diff
@@ -240,8 +240,53 @@ def get_flash_urls(media_url, format_id):
...
+                formats.append({
+                    'url': surl,
+                    'ext': 'm3u8',
+                    'quality': quality
+                })
+        self._sort_formats(formats)
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Sorting is incorrect. Best format is not really the best.

**Comment by einstein95** [comment]
@dstftw Fixes have been applied.

File: youtube_dl/extractor/leeco.py [hunk 8]
```diff
@@ -240,8 +240,56 @@ def get_flash_urls(media_url, format_id):
...
+                    surl = surl['location']
+                except KeyError:
+                    continue
+                formats.append({
+                    'url': surl,
+                    'ext': 'mp4',
+                    'quality': quality
+                })
+        # Commented out due to best quality not being the best
+        # self._sort_formats(formats)
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: This is what you **have to fix** not comment it out.

File: youtube_dl/extractor/leeco.py [hunk 9]
```diff
@@ -240,8 +240,56 @@ def get_flash_urls(media_url, format_id):
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        json = self._download_json('http://player.pc.le.com/player/startup_by_channel_id/1001/%s' % video_id, video_id, query={'host': 'live.le.com'})
+        streams = json['streams']
+        formats = []
+        for stream in streams:
+            surl = stream.get('streamUrl')
+            quality = int(stream.get('rate'))
+            if surl:
+                surl = self._download_json(surl, video_id, query={'format': '1', 'expect': '2', 'platid': '10', 'splatid': '1001', 'uuid': '28394F488CFCF69F405F55B6B046EA4AAC32713C_0', 'vkit': '20161017'})
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Format it properly. This applies for all code.

File: youtube_dl/extractor/leeco.py [hunk 10]
```diff
@@ -240,8 +240,56 @@ def get_flash_urls(media_url, format_id):
...
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        json = self._download_json('http://player.pc.le.com/player/startup_by_channel_id/1001/%s' % video_id, video_id, query={'host': 'live.le.com'})
+        streams = json['streams']
+        formats = []
+        for stream in streams:
+            surl = stream.get('streamUrl')
+            quality = int(stream.get('rate'))
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Breaks if no `rate` key in `stream`.

**Comment by einstein95** [comment]
@dstftw Done

File: youtube_dl/extractor/leeco.py [hunk 11]
```diff
@@ -240,8 +241,65 @@ def get_flash_urls(media_url, format_id):
...
+                    surl = surl['location']
+                except KeyError:
+                    continue
+                formats.append({
+                    'url': surl,
+                    'ext': 'mp4',
+                    'quality': quality
+                })
+
+        formats = sorted(formats, key=lambda k: k['quality'])
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: Use`self. _sort_formats`.

File: youtube_dl/extractor/leeco.py [hunk 12]
```diff
@@ -240,8 +241,65 @@ def get_flash_urls(media_url, format_id):
...
+                     'splatid': '1001',
+                     'uuid': '28394F488CFCF69F405F55B6B046EA4AAC32713C_0',
+                     'vkit': '20161017'
+                     }
+            if surl:
+                surl = self._download_json(surl, video_id, query=query)
+                try:
+                    surl = surl['location']
+                except KeyError:
+                    continue
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: `surl.get(...)` or `try_get`.

File: youtube_dl/extractor/leeco.py [hunk 13]
```diff
@@ -240,8 +241,65 @@ def get_flash_urls(media_url, format_id):
...
+                formats.append({
+                    'url': surl,
+                    'ext': 'mp4',
+                    'quality': quality
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: This is bitrate, not quality.

File: youtube_dl/extractor/leeco.py [hunk 14]
```diff
@@ -240,8 +241,65 @@ def get_flash_urls(media_url, format_id):
...
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        json = self._download_json('http://player.pc.le.com/player/startup_by_channel_id/1001/%s' % video_id,
+                                   video_id, query={'host': 'live.le.com'})
+        streams = json['streams']
+        formats = []
+        for stream in streams:
+            surl = try_get(stream, lambda x: x['streamUrl'])
```
Comment (dstftw) [inline youtube_dl/extractor/leeco.py]: if not surl you should go to the next iteration immediately.

**Comment by dstftw** [comment]
Squash commits.

**Comment by einstein95** [comment]
@dstftw Fixed and squashed

**Comment by yan12125** [comment]
Better to name the extractor LeLunboIE/LeLiveLunboIE/etc. "lunbo" (輪播, alternating videos) are different from ordinary live videos like http://live.le.com/music/play/index.shtml?id=3020170121012628

**Comment by einstein95** [comment]
Updated and squashed

**Pull request closed without merging**